### PR TITLE
feat(portrait): convert component to typescript - FE-4821

### PR DESCRIPTION
**Pull request by mkrds**
### Proposed behaviour

This PR contains Portrait component TS conversion

### Checklist

- [x] Commits follow our style guide
- [x] Related issues linked in commit messages if required
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Portrait component regression

### CODE REVIEW

File: src/components/portrait/portrait-gravatar.component.tsx [hunk 1]
```diff
@@ -4,15 +4,28 @@ import MD5 from "crypto-js/md5";
...
 
 const PortraitGravatar = ({
   gravatarEmail,
   size,
   alt,
   shape,
   errorCallback,
-  ...rest
-}) => {
+}: PortraitGravatarProps) => {
```
Comment (edleeks87) [inline src/components/portrait/portrait-gravatar.component.tsx:28]: You've left the prop types in this file

File: src/components/portrait/portrait.component.tsx [hunk 2]
```diff
@@ -129,60 +177,4 @@ const Portrait = ({
   return renderComponent();
 };
 
-Portrait.propTypes = {
-  ...marginPropTypes,
-  /** The size of the Portrait. */
-  size: PropTypes.oneOf(["XS", "S", "M", "ML", "L", "XL", "XXL"]),
-  /** A custom image URL. */
-  src: (props) => {
```
Comment (edleeks87) [inline src/components/portrait/portrait.component.tsx:137]: Should we have an invariant for the `src` prop? As I don't think any consuming JS projects won't get any feedback that they shouldn't set both `src` and `gravatar` props at the same time

Comment (mkrds) [reply src/components/portrait/portrait.component.tsx:137]: I thought about it and decided that it's not worth it in this case. It's not breaking anything crucial and even if it does it is an immediately visible visual change.

Comment (edleeks87) [reply src/components/portrait/portrait.component.tsx:137]: That's a good point I hadn't checked if it breaks when both are applied, the fact it doesn't means we shouldn't add an invariant 👍

**Review by edleeks87 — APPROVED** [review]

**Comment by grabkowski** [comment]
is there a new feature that is implemented with the refactor? If not, then the commit should be a refactor, as it would not affect developers that are using Carbon.

File: package-lock.json [hunk 3]
```diff
@@ -8661,7 +8667,7 @@
     "ansicolors": {
       "version": "0.3.2",
       "resolved": "https://registry.npmjs.org/ansicolors/-/ansicolors-0.3.2.tgz",
-      "integrity": "sha512-QXu7BPrP29VllRxH8GwB7x5iX5qWKAAMLqKQGWTeLWVlNHNOpVMJ91dsxQAIWXpjuW5wqvxu3Jd/nRjrJ+0pqg==",
+      "integrity": "sha1-ZlWX3oap/+Oqm/vmyuXG6kJrSXk=",
```
Comment (grabkowski) [inline package-lock.json:8670]: is that change expected here? You are only updating the TypeScript types in this PR.

File: package.json [hunk 4]
```diff
@@ -89,6 +89,7 @@
     "@storybook/theming": "^6.4.18",
     "@testing-library/jest-dom": "^5.16.2",
     "@testing-library/react": "^12.1.3",
+    "@types/crypto-js": "^4.1.1",
```
Comment (grabkowski) [inline package.json:92]: have you thought of adding that in a separate commit?

File: src/components/portrait/portrait-initials.component.tsx [hunk 5]
```diff
@@ -1,23 +1,37 @@
...
+  shape = "square",
   darkBackground,
   alt,
-  ...rest
-}) => {
-  const [cachedImageDataUrl, setCachedImageDataUrl] = useState();
+}: PortraitInitialsProps) => {
+  const [cachedImageDataUrl, setCachedImageDataUrl] = useState<string | null>(
+    null
+  );
```
Comment (grabkowski) [inline src/components/portrait/portrait-initials.component.tsx:34]: I think that you could also refactor `setCachedImageDataUrl(null);` to `setCachedImageDataUrl("");`, then you could drop the null type here. IMO as we are refactoring to TypeScript, we should avoid changing types in our variables.

```suggestion
  const [cachedImageDataUrl, setCachedImageDataUrl] = useState<string>();
```

File: package-lock.json (hunk 3 above)
Comment (mkrds) [reply package-lock.json:8670]: Well, that comes from installation of `@types/crypto-js` package

**Review by edleeks87 — DISMISSED** [review]

File: package-lock.json (hunk 3 above)
Comment (grabkowski) [reply package-lock.json:8670]: but that should be done by `crypto-js` package, and this package contains just TypeScript types for it

**Review by edleeks87 — APPROVED** [review]

**Review by grabkowski — APPROVED** [review]

**Review by ZhuoyuJin — APPROVED** [review]

**Pull request merged**